### PR TITLE
API Call for Promos

### PR DESCRIPTION
**Pull request by arpavlic03**
Backend for the Promo Form to post into AirTable - 10TVPromos.

### CODE REVIEW

File: _includes/promos/10tvpromo.html [hunk 1]
```diff
@@ -59,8 +59,14 @@ <h4><br/>Offer valid until May 1st, 2020<br />To submit your business, fill out
   </div>
   
 
+  <script  src="https://code.jquery.com/jquery-3.2.1.slim.min.js"
```
Comment (maharris1011) [inline _includes/promos/10tvpromo.html:62]: I think this is not necessary. We load it in the footer.

File: _includes/promos/10tvpromo.html [hunk 2]
```diff
@@ -90,14 +96,36 @@ <h4><br/>Offer valid until May 1st, 2020<br />To submit your business, fill out
...
+                                e.preventDefault();
+                                console.log("FORM NOT VALID");
                             } else {
-                            submitText.style.color = 'black';
-                            submitButton.className = 'btn btn-success';
-                            submitButton.innerHTML = '&check;';
-                            submitButton.disabled = true;
+                                e.preventDefault();
+                                var xhr = new XMLHttpRequest();
+                                var myAPIKey = process.env.AIRTABLE_KEY;
```
Comment (maharris1011) [inline _includes/promos/10tvpromo.html:107]: Where is the environment variable set? This code will execute on the client who likely doesn’t know our Airtable key.

**Comment by arpavlic03** [comment]
Hi, once I get back online, I will change the code to use site.env.API_KEY as a hidden element in the template...which is already configured in the jekyll configuration. That way the client can obtain the api key to use to post to the airtable

**Comment by maharris1011** [comment]
i'm not sure about that either. That'll work, but the challenge is that it'll generate directly into the site code. The reason it is OK now is that the API key never shows up in the HTML/JS that goes to the client. It's only used at build time by a plugin. That won't work in this case, b/c you need to write to the Airtable directly and to do that the javascript has to have the key. The user could always "View Source" in the browser and see the key.

I think opening up an API on Amazon Lambda or Google Functions or Firebase is the right way to go in this case. With the API we could sanitize the input, do some business logic (like sending email), and protect the API key. 

I'm super happy to work on this with someone if they want to. Like I mentioned, @CoreySchnedl did a killer job with some research into Firebase functions as APIs the other day.

**Comment by CoreySchnedl** [comment]
@maharris1011 @arpavlic03 do we want to set up a quick call or slack thread to discuss what the best solution for this is? I am available anytime this evening. Unfortunately, there's no real way to do this 100% securely from what I've seen. Either A, we expose the api key client side and bad users can just make the same api calls with the api key, or b we move the logic to some back end, but that back end will be just as vulnerable option A. Moving to lambda would only move the vulnerability to someone trying to take advantage of the public aws api gateway endpoint, and would only complicate things even if our api key would technically remain secret.. IMHO, although we probably don't want people having to do manual tasks, the best way to get around this problem is to change that form into an email template that goes to a new email that we don't care if malicious users try to spam it. Or, if that doesn't seem best, we could just add a google recaptcha to at least protect from completely malicious attacks. https://developers.google.com/recaptcha.

**Comment by arpavlic03** [comment]
Hello, unfortunately tonight is not the best night for me for a discussion. Anytime tomorrow will be better day for me to talk about this in detail

**Comment by maharris1011** [comment]
@arpavlic03 - i opened up an api this weekend at https://wduc7ys73l.execute-api.us-east-1.amazonaws.com/dev/promotions/create that takes form input & adds it to the Promotions view of the Airtable base, using the same key that @jranta used in his PRs.

if you modify your form to use `action="https://wduc7ys73l.execute-api.us-east-1.amazonaws.com/dev/promotions/create"` `method="POST"` and change the "name" attributes on the form items to "Business", "Promo", and "Website", respectively, you don't need any javascript or api keys or anything on the page for posting to Airtable.

It solves the problems with api keys we talked about and reduces the need for code in your PR.

Attached is a Postman file that you can import to work with the API directly (change to .json extension first)
[CantStopColumbus.postman_collection.txt](https://github.com/maharris1011/cantstopcolumbus/files/4438731/CantStopColumbus.postman_collection.txt)

File: _includes/promos/10tvpromo.html [hunk 3]
```diff
@@ -73,6 +85,36 @@ <h4><br/>Offer valid until May 1st, 2020<br />To submit your business, fill out
...
+         console.log(body["records"] != []);
+         console.log(body["records"].length > 0);
+            
+         if(body["records"].length <= 0)
+            resolve("Success");
+         else
+            reject("Failure");
+        }; });
+    
+        async function myFunction() {
```
Comment (seanlawrenz) [inline _includes/promos/10tvpromo.html:112]: If we want to use async we'll need to add babel to the project

File: _includes/promos/10tvpromo.html [hunk 4]
```diff
@@ -81,26 +123,81 @@ <h4><br/>Offer valid until May 1st, 2020<br />To submit your business, fill out
...
+                                form.classList.add('was-validated');
                             } else {
-                            submitText.style.color = 'black';
-                            submitButton.className = 'btn btn-success';
-                            submitButton.innerHTML = '&check;';
-                            submitButton.disabled = true;
+                                e.preventDefault();
+                                console.log("COMMENCING CHECK");
+                                myFunction()
+                                .catch(stuff => {
```
Comment (seanlawrenz) [inline _includes/promos/10tvpromo.html:147]: mismatching .catch and `async`.

On async and await functions, you want to use a pattern of `try ... catch`.

Comment (arpavlic03) [reply _includes/promos/10tvpromo.html:147]: Hi Sean! I will be talking to Mark tomorrow but there might be a possibility that we don't need the async/await at all if we call a lambda-oriented api that can handle the post for us and hopefully check for duplicate records. 

In any other case, I would definitely agree with you. It worked locally on my machine with the catch, .then, asyncs, awaits and Promises with good ol' jquery but instilling try/catch in all of my functionality will be much more helpful.

**Comment by arpavlic03** [comment]
@maharris1011, just wanted to let you know before our call that the api works! Thank you so much for all of your hard work with this.

**Pull request closed without merging**